Allow users to update Upgrade Window Start

mcbattirola opened the pull request:
This PR adds a prop `children` to the "Help & Support" page that allows enterprise to extend it. `webapps.e` will have its own page that will add a new section to Support with some cloud specific information. Some components of `Support` have been made more generic and exported so we can use them on enterprise.

This PR also updates the events table to properly show the new audit event type.

Enterprise PR using this feature: https://github.com/gravitational/webapps.e/pull/357

## Testing

To run the whole thing, you can:

1. Run salescenter locally (follow [this](https://github.com/gravitational/cloud#sales-center))
2. Checkout `master` ~`mcbattirola/add-cloud-upgrade-windows`~ on both `teleport` and `teleport.e`, and `webapps.e` on branch `mcbattirola/add-cloud-upgrade-windows`
3. comment [this line](https://github.com/gravitational/teleport/blob/master/lib/modules/modules.go#L130) on `lib/modules/modules.go` so we can run cloud without second factor
4. build and run **teleport.e** with [this](https://github.com/gravitational/cloud/blob/master/db/fixtures/license.pem) license (under auth_service in the config file). I recommend you copy `cloud/dev/teleport/config.yaml` and add:

```yaml
auth_service:
  enabled: "yes"
  listen_addr: 0.0.0.0:3025
  tokens:
  - proxy,node:[token]
  authentication:
  license_file: /home/matheus/tc/db/fixtures/license.pem
```

You have to run it like this:

```bash
TELEPORT_CLOUD_HOSTPORT="api.localhost:3080" DEBUG=1 /path/to/teleport/e/build/teleport start -c my-config-file.yaml -d --insecure
```

6. Run this branch of the teleport UI with `yarn start-teleport-e --target=https://localhost:4080/web`
Then you should be able to make a request from the local teleport to the salescenter

## Screenshots

![Screenshot from 2022-07-14 12-23-57](https://user-images.githubusercontent.com/8400114/179018581-4a36bd8e-9a7d-4b9d-88a0-db90daf9ed23.png)


![Screenshot from 2022-07-19 14-00-08](https://user-images.githubusercontent.com/8400114/179807704-aa25cc24-9490-455e-8e8d-88f05be06f3a.png)

Closes https://github.com/gravitational/cloud/issues/1808

kimlisa commented on packages/teleport/src/config.ts:
```diff
@@ -55,6 +56,12 @@ const cfg = {
 
   proxyCluster: 'localhost',
 
+  scheduledUpgradeWindows: [
+    '08:00:00',
+    '16:00:00',
+    '23:00:00',
+  ] as UpgradeWindowStart[],
```
im not sure this is the right place, these are just consts and i would move and define these consts to the component that it is being used

kimlisa commented on packages/teleport/src/services/upgradeWindow/upgradeWindow.ts:
```diff
@@ -0,0 +1,40 @@
+  },
+};
+
+export default service;
```
we are moving away from default export, btw you can have a `type` and a `var` be same name i believe

kimlisa commented on packages/teleport/src/Support/ScheduleUpgrades/ScheduleUpgrades.tsx:
```diff
@@ -0,0 +1,108 @@
+  onSelectedWindowChange,
+  attempt,
+}: Props) {
+  const options =
+    attempt.status === 'processing'
+      ? []
+      : upgradeWindowOptions.map((window: UpgradeWindowStart) => ({
+          label: makeLabel(window),
+          value: window,
+        }));
```
```suggestion
  let options = []
  if (attempt.status === 'success') {
    options = upgradeWindowOptions.map((window: UpgradeWindowStart) => ({
          label: makeLabel(window),
          value: window,
        }));
  }
```
easier to read imo. also i think we are over complicating this. the update window times are just consts, which is nice, then we can just move this outside of the component:

```ts
const upgradeWindowOptions = [...]

...
   const options = attempt.status === 'processing' ? [] : upgradeWindowOptions

```

mcbattirola replied:
Yes, this is easier to read, thanks!

I'd like to keep the options being a result of `.map()` so we have a distinction between the option and the actual value (easier to track since the value itself will be a state in the hook).

What doo you think about doing something like this?

```jsx
const availableUpgradeWindowStarts: UpgradeWindowStart[] = [
  '08:00:00',
  '16:00:00',
  '23:00:00',
];

const upgradeWindowOptions = availableUpgradeWindowStarts.map(
  (window: UpgradeWindowStart) => ({
    label: makeLabel(window),
    value: window,
  })
);

export function ScheduleUpgrades({
  const options = attempt.status === 'processing' ? [] : upgradeWindowOptions
// ...
```

I implemented this in https://github.com/gravitational/webapps/pull/980/commits/58e4996d1bee3d941bce01650930183f354e08e2, please let me know what you think

kimlisa commented on packages/teleport/src/Support/ScheduleUpgrades/ScheduleUpgrades.tsx:
```diff
@@ -0,0 +1,108 @@
+      open={true}
+      dialogCss={() => ({ maxWidth: '600px' })}
+    >
+      <DialogTitle mr="auto">Select a maintenance start window</DialogTitle>
```
i think this is usually wrapped in a `DialogHeader` and you won't need that `mr` styling (look at User components like `UserAddEdit`)

kimlisa commented on packages/teleport/src/Support/ScheduleUpgrades/ScheduleUpgrades.tsx:
```diff
@@ -0,0 +1,108 @@
+          Teleport provides different cluster maintenance windows to minimize
+          downtime during upgrades, patches, etc.
+        </Text>
+        {attempt.status === 'failed' && (
+          <Alert kind="danger" children={attempt.statusText} />
+        )}
```
most errors usually are displayed right underneath the title (there are exceptions but this one isn't it, look at some of the User components, very similar styling)

kimlisa commented on packages/teleport/src/Support/ScheduleUpgrades/ScheduleUpgrades.tsx:
```diff
@@ -0,0 +1,108 @@
+          onChange={handleChange}
+          menuPosition="fixed"
+          options={options}
+          value={
+            attempt.status === 'processing'
+              ? {
+                  label: 'Processing...',
+                  value: null,
+                }
+              : { value: selectedWindow, label: makeLabel(selectedWindow) }
+          }
```
nice touch, but i would just disable it (this is the current pattern)

kimlisa commented on packages/teleport/src/Support/ScheduleUpgrades/ScheduleUpgrades.tsx:
```diff
@@ -0,0 +1,108 @@
+          onClick={onClose}
+          disabled={attempt.status === 'processing'}
+        >
+          Close
```
i believe the pattern is `cancel` instead of `close`

kimlisa commented on packages/teleport/src/Support/ScheduleUpgrades/ScheduleUpgrades.tsx:
```diff
@@ -0,0 +1,108 @@
+export type Props = {
+  onClose: () => void;
+  onSave: (window: UpgradeWindowStart) => any;
+  upgradeWindowOptions: string[];
```
i would move `upgradeWindowOptions` and their consts to this file at top

kimlisa commented on packages/teleport/src/Support/useUpgradeWindowStart.ts:
```diff
@@ -0,0 +1,69 @@
+      ctx.upgradeWindowService
+        .getUpgradeWindowStart()
+        .then(setSelectedUpgradeWindowStart)
+    ).then(success => {
+      if (!success) {
+        setSelectedUpgradeWindowStart(upgradeWindowOptions[0]);
+      }
+    });
```
i'm not sure i understand when do we get `success`? (or rather what is it?)

mcbattirola replied:
`success` is the boolean returned by `run()` after a request. But I guess this isn't necessary since we can just handle `attempt.status === 'failed'`

kimlisa replied:
i don't see this code anymore right?

mcbattirola replied:
Yes, sorry. I removed it.

kimlisa commented on packages/teleport/src/Support/useUpgradeWindowStart.ts:
```diff
@@ -0,0 +1,69 @@
+    return run(() =>
+      ctx.upgradeWindowService
+        .updateUpgradeWindowStart(selectedUpgradeWindowStart)
+        .then(() => hideScheduleUpgrade())
```
```suggestion
        .then(() => closeScheduleUpgrade())
```

kimlisa commented on packages/teleport/src/Support/useUpgradeWindowStart.ts:
```diff
@@ -0,0 +1,69 @@
+    useState<UpgradeWindowStart>();
+
+  useEffect(() => {
+    run(() =>
+      ctx.upgradeWindowService
+        .getUpgradeWindowStart()
+        .then(setSelectedUpgradeWindowStart)
+    ).then(success => {
```
wouldn't this run for non cloud versions?

kimlisa commented on packages/teleport/src/Support/Support.tsx:
```diff
@@ -102,29 +115,61 @@ export const Support = ({
+          />
+          <Text>
+            Window Start Time is the hour in which an upgrade may begin.
+            Changing this value changes it for everyone in your org.
```
```suggestion
            Changing this value changes it for everyone in your organization.
```

kimlisa commented on packages/teleport/src/Support/Support.tsx:
```diff
@@ -102,29 +115,61 @@ export const Support = ({
+            data={
+              <span>
+                {selectedUpgradeWindowStart}
+                <Edit onClick={showScheduleUpgrade} ml="2" />
```
anything clickable should have `cursor: pointer`, and i believe they should also have like a hover state so when you hover over it it like changes color like most buttons do

kimlisa commented on packages/teleport/src/services/upgradeWindow/upgradeWindow.ts:
```diff
@@ -0,0 +1,40 @@
+limitations under the License.
+*/
+
+import cfg from 'e-teleport/config';
```
we can't import things from `e-teleport` in `teleport` b/c `e-teleport` is not a dependency. so we either move the enterprise URL endpoint to OS repo (which i'm not sure if that's okay to do), OR enterprise will need to have it's own `Support` component (which is okay to duplicate)

I would ask for advice on how to proceed on this.  i've seen components handling both OS and enterprise, and if it's trivial enough, it can be in OS, but i also know that we want more things to be in enterprise as well so I'm unsure how to proceed :woman_shrugging: maybe @zmb3 ?

mcbattirola replied:
Hmm right, thanks.

I think we should keep the URL endpoint in the enterprise.

We can add a `children` props to `Support`, then `webapps.e` has its own `Support` that uses OSS:

SupportE:

```jsx
// packages/webapps.e/teleport/src/Support/Support.tsx
export default function Container(): JSX.Element {
  const ctx = useTeleport();
  // const cluster = ctx.storeUser.state.cluster;
  const upgradeWindowsState = useUpgradeWindowStart(ctx);

  return <SupportE {...upgradeWindowsState} isCloud={cfg.isCloud} />;
}

export const SupportE = ({
  isCloud,
  showScheduleUpgrade,
  scheduleUpgradesVisible,
  closeScheduleUpgrade,
  onUpdate,
  selectedUpgradeWindowStart,
  setSelectedUpgradeWindowStart,
  attempt,
  ...props
}: Props) => {
  return (
    <>
      <Support {...props}>
        {isCloud && (
          <DataContainer title="Scheduled Upgrades">
            <DataItem
              title="Window Start Time"
              data={
                <span>
                  {selectedUpgradeWindowStart}
                  <EditLink onClick={showScheduleUpgrade} ml="2" />
                </span>
              }
            />
            <Text>
              Window Start Time is the hour in which an upgrade may begin.
              Changing this value changes it for everyone in your organization.
            </Text>
          </DataContainer>
        )}
        {scheduleUpgradesVisible && (
          <ScheduleUpgrades
            onSave={onUpdate}
            onCancel={closeScheduleUpgrade}
            selectedWindow={selectedUpgradeWindowStart}
            onSelectedWindowChange={setSelectedUpgradeWindowStart}
            attempt={attempt}
          />
        )}
      </Support>
    </>
  );
};
```

Support OSS:

```jsx
// packages/teleport/src/Support/Support.tsx
export default function Container({ children }): JSX.Element {
  const ctx = useTeleport();
  const cluster = ctx.storeUser.state.cluster;
  // no more upgrade window stuff
  return (
    <Support
      {...cluster}
      isEnterprise={cfg.isEnterprise}
      tunnelPublicAddress={cfg.tunnelPublicAddress}
      isCloud={cfg.isCloud}
      children={children}
    />
  );
}
export const Support = ({
  clusterId,
  authVersion,
  publicURL,
  isEnterprise,
  tunnelPublicAddress,
  isCloud,
  children,
}: Props) => {
// ...
return (
    <FeatureBox pt="4">
      ...
      {children}
    </FeatureBox>
```

This way all the upgrade window stuff is moved to enterprise, and we don't have to duplicate code. The downside is that is kind of strange to have `Support` receiving children, it can be confusing.

kimlisa replied:
i think this is fine to me, we did something similar before but i don't remember which component

mcbattirola commented on packages/design/src/Icon/Icon.jsx:
```diff
@@ -111,6 +111,7 @@ export const ClipboardUser = makeFontIcon(
   'ClipboardUser',
   'icon-clipboard-user'
 );
+export const Clock = makeFontIcon('Clock', 'icon-clock3');
```
Yes, I updated the icomoon lib when started working on this issue and since then the assets were updated on https://github.com/gravitational/webapps/pull/952/files#diff-cf3b0c544bd0e10329bf5ab41682df0c6156bfe9bee8096ef83ff7b0f865765e, that's why I didn't had to update the assets in this PR.

kimlisa replied:
i think it's ok to keep it

kimlisa commented on packages/teleport/src/Support/ScheduleUpgrades/ScheduleUpgrades.tsx:
```diff
@@ -0,0 +1,110 @@
+  DialogHeader,
+  DialogTitle,
+} from 'design/Dialog';
+import Select, { Option } from 'shared/components/Select';
+import { Attempt } from 'shared/hooks/useAttemptNext';
+import { UpgradeWindowStart } from 'teleport/services/upgradeWindow/upgradeWindow';
+import Text from 'design/Text';
+import Alert from 'design/Alert';
```
i haven't been paying attention to these, but just like how teleport has import ordering, webapps has its own: https://gravitational.slack.com/archives/C02DQ1C2BMW/p1643999171617869

zmb3 replied:
I'm becoming less okay with us just linking to a random slack thread from 5 months ago and expecting that people should have just known it existed.

Import order is largely a matter of personal preference and doesn't affect functionality in any way.

For things like this (no impact on behavior and hard to get everyone to agree on) we need to decide whether or not we care.

- If we care: we need to enforce it with tooling and not tribal knowledge so that devs don't ever have to think about it.
- Otherwise: we don't care and devs should simply not have to think about it.

kimlisa commented on packages/teleport/src/Support/ScheduleUpgrades/ScheduleUpgrades.tsx:
```diff
@@ -0,0 +1,110 @@
+import Text from 'design/Text';
+import Alert from 'design/Alert';
+
+const availableUpgradeWindowStarts: UpgradeWindowStart[] = [
+  '08:00:00',
+  '16:00:00',
+  '23:00:00',
+];
```
something about this was bugging me and i just realized it! it was weird to be defining these consts twice. i think this is what you wanted, we can derive the type from the array:

```ts
type UpgradeWindowStart = typeof availableUpgradeWindowStarts[number];
// you get:  "08:00:00" | "16:00:00" | "23:00:00"
```

so i think these definitions (array and type) actually belong in the service file?

mcbattirola replied:
Awesome, this is what I wanted!
Btw I had to add `as const` after `availableUpgradeWindowStarts`, otherwise I get `type UpgradeWindowStart = string`:

```js
const availableUpgradeWindowStarts = ['08:00:00', '16:00:00', '23:00:00'];
export type UpgradeWindowStart = typeof availableUpgradeWindowStarts[number];
// type UpgradeWindowStart = string

const availableUpgradeWindowStarts = ['08:00:00', '16:00:00', '23:00:00'] as const;
export type UpgradeWindowStart = typeof availableUpgradeWindowStarts[number];
//type UpgradeWindowStart = "08:00:00" | "16:00:00" | "23:00:00"

```

kimlisa commented on packages/teleport/src/Support/ScheduleUpgrades/ScheduleUpgrades.tsx:
```diff
@@ -0,0 +1,110 @@
+          Teleport provides different cluster maintenance windows to minimize
+          downtime during upgrades, patches, etc.
+        </Text>
+        <Select
+          onChange={handleChange}
+          menuPosition="fixed"
+          options={options}
+          isDisabled={attempt.status === 'processing'}
+          value={{ value: selectedWindow, label: makeLabel(selectedWindow) }}
+        ></Select>
```
```suggestion
        <Select
          onChange={handleChange}
          menuPosition="fixed"
          options={options}
          isDisabled={attempt.status === 'processing'}
          value={{ value: selectedWindow, label: makeLabel(selectedWindow) }}
        />
```

kimlisa commented on packages/teleport/src/Support/useUpgradeWindowStart.ts:
```diff
@@ -0,0 +1,62 @@
+    return run(() =>
+      ctx.upgradeWindowService
+        .updateUpgradeWindowStart(selectedUpgradeWindowStart)
+        .then(() => closeScheduleUpgrade())
```
should be same

```suggestion
        .then(closeScheduleUpgrade)
```

kimlisa commented on packages/teleport/src/Audit/EventList/EventTypeCell.tsx:
```diff
@@ -128,6 +128,7 @@ const EventIconMap: Record<EventCode, React.FC> = {
   [eventCodes.X11_FORWARD]: Icons.Info,
   [eventCodes.X11_FORWARD_FAILURE]: Icons.Info,
   [eventCodes.CERTIFICATE_CREATED]: Icons.Keypair,
+  [eventCodes.UPGRADE_WINDOW_UPDATED]: Icons.Info,
```
yes, a doc on our README will be greatly appreciated! refer to this comment https://github.com/gravitational/webapps/pull/970#pullrequestreview-1037970565 (not sure what the status of the PR is, but if you get to it first i'd mark it of completed)

kimlisa commented on packages/teleport/src/Support/ScheduleUpgrades/ScheduleUpgrades.story.tsx:
```diff
@@ -0,0 +1,61 @@
+  attempt: {
+    status: '',
+    statusText: '',
+  } as any,
```
is the `as any` required? we might have some old code which naively uses them, i'm guilty of it :)

kimlisa commented on packages/teleport/src/Support/ScheduleUpgrades/ScheduleUpgrades.tsx:
```diff
@@ -0,0 +1,110 @@
+}: Props) {
+  const options = attempt.status === 'processing' ? [] : upgradeWindowOptions;
+
+  const handleChange = (selected: Option<any>) =>
```
think you meant this: 
```suggestion
  const handleChange = (selected: Option<UpgradeWindowStart>) =>
```

kimlisa commented on packages/teleport/src/Support/ScheduleUpgrades/ScheduleUpgrades.tsx:
```diff
@@ -0,0 +1,110 @@
+}
+
+export type Props = {
+  onCancel: () => void;
```
we prefer this style of defining funcs
```suggestion
  onCancel(): void;
```

kimlisa commented on packages/teleport/src/Support/ScheduleUpgrades/ScheduleUpgrades.tsx:
```diff
@@ -0,0 +1,110 @@
+
+export type Props = {
+  onCancel: () => void;
+  onSave: (window: UpgradeWindowStart) => any;
```
i think you meant `void`?

kimlisa reviewed (commented):
let's include the changes for enterprise that you proposed and i will do a final round

mcbattirola commented:
Thank you @kimlisa !

I moved the upgrade window stuff to enterprise ([PR here](https://github.com/gravitational/webapps.e/pull/357)), now this PR only adds the event and adds a children prop to Support. I've updated the PR text to reflect this

kimlisa commented on packages/teleport/src/Support/Support.tsx:
```diff
@@ -15,14 +15,14 @@
-import styled from 'styled-components';
 
-export default function Container() {
+export default function Container({ children }): JSX.Element {
```
is that children really typed? it looks like only the return value is typed but that should be inferred?

ryanclark commented on packages/teleport/src/Support/Support.tsx:
```diff
@@ -15,14 +15,14 @@
-import styled from 'styled-components';
 
-export default function Container() {
+export default function Container({ children }: { children?: JSX.Element }) {
```
Can you use `React.ReactNode` here? It's what the official typings use for any React renderable element - https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/react/index.d.ts#L799

mcbattirola replied:
Thanks

ryanclark commented on packages/teleport/src/Support/Support.tsx:
```diff
@@ -215,11 +226,12 @@ const Header = ({ title = '', icon = null }) => (
   isEnterprise: boolean;
   isCloud: boolean;
   tunnelPublicAddress?: string;
+  children?: JSX.Element;
```
```suggestion
  children?: React.ReactNode;
```

ryanclark commented on packages/teleport/src/Support/Support.tsx:
```diff
@@ -15,14 +15,14 @@
-import styled from 'styled-components';
 
-export default function Container() {
+export default function Container({ children }: { children?: JSX.Element }) {
```
```suggestion
export default function Container({ children }: { children?: React.ReactNode }) {
```

mcbattirola commented:
@ryanclark Thanks for the review, are you ok with merging this?

ryanclark reviewed (approved):
LGTM!
